README.md

```diff
@@ -5,3 +5,134 @@
+instead had pretty much any kind of neural network, Gaussian process, ODE, or really just
+any model with more than a couple of distinct parameters. From the perspective of
+writing complicated models, implementing things in terms of a single vector of
+parameters that is _manually_ chopped up is an _extremely_ bad design choice. It simply
+doesn't scale.
+
+However, a single vector of e.g. `Float64`s _is_ extremely convenient when writing general
+purpose optimisers / approximate inference routines --
+[Optim.jl](https://github.com/JuliaNLSolvers/Optim.jl) and
+[AdvancedHMC.jl](https://github.com/TuringLang/AdvancedHMC.jl) being two obvious examples.
+
+
+
+
+
+# The ParameterHandling.jl Approach
+
+`ParameterHandling.jl` aims to give you the best of both worlds by providing the tools
+required to automate the transformation between a "structured" representation (e.g. nested
+`NamedTuple` / `Dict` etc) and a "flattened" (e.g. `Vector{Float64}`) of your model
+parameters.
+
+The function `flatten` eats a structured representation of some parameters, returning the
+flattened representation _and_ a function that converts the flattened thing back into its
+structured representation.
+
+`flatten` is implemented recursively, with a _very_ small number of base-implementations
+that don't themselves call `flatten`.
+
+You should expect to occassionally have to extend `flatten` to handle your own types and, if
+you wind up doing this for a function in `Base` that this package doesn't yet cover, a PR
+including that implementation will be very welcome.
+
+See `test/parameters.jl` for a couple of examples that utilise `flatten` to do something
+similar to the task described above.
+
+
+
+
+
+# Dealing with Constrained Parameters
+
+It is very common to need to handle constraints on parameters e.g. it may be necessary for a
+particular scalar to always be positive. While `flatten` is great for changing between
+representations of your parameters, it doesn't really have anything to say about this
+constraint problem.
+
+For this we introduce a collection of new `AbstractParameter` types (whether we really need
+them to have some mutual supertype is unclear at present) that play nicely with `flatten`
+and allow one to specify that e.g. a particular scalar must remain positive, or should be
+fixed across iterations. See `src/parameters.jl` and `test/parameters.jl` for more examples.
+
+The approach to implementing these types typically revolves around some kind of deferred /
+delayed computation. For example, a `Positive` parameter is represented by an
+"unconstrained" number, and a "transform" that maps from the entire real line to the
+positive half. The `value` of a `Positive` is given by the application of this transform to
+the unconstrained number. `flatten`ing a `Positive` yields a length-1 vector containing the
+_unconstrained_ number, rather than the value represented by the `Positive` object.
```

Comment:
I think I understand the need for `AbstractParameter`s but am not certain how a user would interact with them from this description.  Would they pass `AbstractParameter`s to `flatten`?  Or would their parameters be abstracted during `flatten`?

Also, is it likely that users would want to add to the collection of `AbstractParameter`s as is the case with extending `flatten` to handle more types?

Reply:
I've added a worked example and a note regarding extension. Could you see if it's clear now?

Reply:
The example makes things clearer.